### ISSUE TITLE
Edge in backtest

### ISSUE THREAD

**Issue by Fimap9to1** [freqtrade/freqtrade#5095]
Hi Matthias, I know that the edge function in backtesting is already been discussed in issue #4826 and that, as you've written, it will take more time to backtest results if this were to be simulated properly.

But, I was thinking, if the most effective action I can take to improve my strategy is to work on stoploss, what can I do in order to get a reliable backtest?
Custom stoploss has proven to be unreliable in this regard (backtest and live often differ sensitively). The only really reliable operations seems to be orders executed on close (sell/buy) and at a fixed price (normal stoploss).

Since custom stoploss seems a jungle to cross, if we were to "change its behaviour" to a common stoploss behaviour. Why don't enable the edge positioning?

Like you did with protections, selectively activate it in backtests if one needs it. What do you think? Can we talk about how to do it? 

Thanks for reading!

**Comment by xmatthias** [comment]
Custom stoploss (and trailing stop too, for that matter) are unreliable with long timeframes, as the "black spot" intra-candle is becomming long. 

Backtesting with short timeframes (5m / 15m max) will give you shorter "black spots" intra-candle - and should therefore improve accuracy - obviously always if you stay within sane limits (don't use stoploss below the average candle movements).

---

If you need edge to run on historic data, use `freqtrade edge`.

You can sure try to activate edge in backtesting - but i'll not be supporting in this endavor (also not by answering questions) as i never thought about it (so i don't know how to approach that) - but also have no interrest in it as i don't see the benefit (and i don't see what it would bring you on top of custom-stoploss if that is done right).

As such, i see my time spent better elsewhere.
If you intend to dvelve into that and get as far as making a PR, i will/can-not guarantee merging / accepting of it as i'm not sure what it would involve - and by accepting a PR, i'm explicitly accepting to support / maintain that going forward.

**Comment by Fimap9to1** [comment]
Thank you for your answer!

I was hoping to make my point clearer, maybe I mislead you.

Edge is a great idea in my opinion. So great that I think it deserves a deeper development. If we were able to use it in backtesting it will be very good for:

1) Optimizing win/loss ratio;
2) Dinamically set stoplosses;
3) Get a more reliable backtests with a dinamic stoploss even in larger timeframes (as in short tf as well obviously);
4) Further understand how a specific pair behaves and from that improve performance of it selectively
5) Get to know how total profit, drawdown, etc, change using it or not using it (the edge table does not give this information at the moment, maybe just adding a total profit column and a drawdown column to it is possible? Let me know what you think ;) )

About Custom Stoploss, it is usefull for those who have a trend following strategy. Although, unreliable for those who have a reversal strategy. Since many times the reversal is a fake reversal, therefore price goes up a little and after that it dumps again (so is very unreliable to assume that high happen first in these cases). I use the 1h tf and it still works unreliably because of the way he works (as described in the assumptions) and the kind of strategy adopted.

For these reasons I think it would be very usefull and appreciated by many. Did I convinced you? ;)

Thank you in advance for your kind response!

**Comment by xmatthias** [comment]
You can still use custom stoploss to set an ["absolute" stoploss](https://www.freqtrade.io/en/latest/strategy-advanced/#calculating-stoploss-relative-to-open-price) (without trailing functionality (simply never change the value relaive to the open price) - based on some arbitrary parameters - so i think your proposal is based on not understanding / properly reading the docs on how custom_stoploss works and instead trying to go another, much more complex, route.

---

there's little reason to convince me - nothing can prevent you from working on it - but as mentioned above - i'll not be spending my time on it (also not assisting / answering questions on how to do it - at least not before a PR is "review ready").

It'll have a pretty huge impact (code / performance wise) on backtesting if done right (you'll need to run edge every few candles with a limited lookback to simulate how it would work "in real mode" to get reliable results)  - and as such, it'll become very slow.
As i also expect the code to become rather complex - for now, i have my doubts that i'll be accepting / merging a Pull request (as that will mean i'll accept responsability to maintain it should it break in the future).

Obviously, you can still do a PR and I'll guarantee i'll have a look / review that - and if it's done well, has no significant impact on code complexity, I'll fo sure consider it - but i still think it's most likely wasted time better spent elsewhere.

**Comment by Fimap9to1** [comment]
I always read the docs carefully before writing, but If I need to calculate the stoploss based on the environment (indicators value at buy candle or stop calculation based on previuous stoploss behaviuor) coin by coin? This is not possible with the "absolute" stoploss and it doesn't work with this method in backtest (this method: https://github.com/freqtrade/freqtrade-strategies/blob/master/user_data/strategies/fixed_riskreward_loss.py).

Since custom stoploss is tied to current price, and "absolute" stoploss is not able to use indicators as far as I understand.

What else can I try? I'd like to set a stoploss from an indicator's value at buy candle and let it there until it is hit. For example: stoploss = buy_price - (atr_at_buy_price * 2.5), that's it. No trailing, offset or other functionalities.

And not using the trailing assumption, but the stoploss assumption.

Thank you for you patience! It is very welcome nowadays!

**Comment by xmatthias** [comment]
Simply combine all the information available on the [advanced strategy page](https://www.freqtrade.io/en/latest/strategy-advanced/).

You can get [access to the dataframe](https://www.freqtrade.io/en/latest/strategy-advanced/#dataframe-access) (at current price, or buy price) - and then set the stoploss [relative to trade open](https://www.freqtrade.io/en/latest/strategy-advanced/#calculating-stoploss-relative-to-open-price) instead of returning the value directly.

Now i'll surely not code that up for you - but all the information is available and can (more or less) be copy/pasted together if you think about it carefully.

**Comment by Fimap9to1** [comment]
Ok thanks for your kindness! Have a nice time!
